edx_sga/tests/integration_tests.py

```diff
@@ -801,3 +807,91 @@ def test_grace_period(self, has_grace_period):
+            </vertical>""".format(attrs=attrs)
+        ) if solution_as_element else (
+            """<vertical display_name="SGA Unit">
+                <edx_sga url_name="edx_sga" xblock-family="xblock.v1" display_name="SGA Test 1" {attrs} />
+            </vertical>""".format(attrs=attrs)
+        )
+
+    def import_test_course(self, solution_as_attribute, solution_as_element):
+        """
+        Import the test course with the sga unit
+        """
+        # adapted from edx-platform/cms/djangoapps/contentstore/management/commands/tests/test_cleanup_assets.py
+        root = os.path.dirname(os.path.dirname(os.path.dirname(__file__)))
+        xml_dir = os.path.join(root, "test_data")
+
+        with open(os.path.join(xml_dir, "2017_SGA", "vertical", "vertical.xml"), "w") as f:
+            f.write(self.make_test_vertical(solution_as_attribute, solution_as_element))
+
+        store = modulestore()
+        import_course_from_xml(
+            store,
+            'sga_user',
+            xml_dir,
+        )
+
+        return store.get_course(CourseLocator.from_string('SGAU/SGA101/course'))
+
+    @data(*[
+        [True, True],
+        [True, False],
+        [False, True],
+        [False, False],
+    ])
+    @unpack
+    def test_import(self, solution_as_attribute, solution_as_element):
+        """Import the test course with the SGA module"""
+        course = self.import_test_course(solution_as_attribute, solution_as_element)
+        sga = course.get_children()[0].get_children()[0].get_children()[0].get_children()[0]
+
+        if solution_as_attribute:
+            expected_solution = "<p>Broken xml"
+        elif solution_as_element:
+            expected_solution = "<p>You're seeing the answer</p>"
+        else:
+            expected_solution = ''
+        assert sga.solution == expected_solution
+        assert sga.showanswer == ShowAnswer.PAST_DUE
+
+    @data(*[
+        [True, True],
+        [True, False],
+        [False, True],
+        [False, False],
+    ])
+    @unpack
+    def test_export(self, solution_as_attribute, solution_as_element):
+        """Export the test course with the SGA module"""
+        course = self.import_test_course(solution_as_attribute, solution_as_element)
+
+        temp_dir = tempfile.mkdtemp()
+        self.addCleanup(lambda: shutil.rmtree(temp_dir))
+
+        store = modulestore()
+        export_course_to_xml(store, None, course.id, temp_dir, "2017_SGA")
+
+        with open(os.path.join(temp_dir, "2017_SGA", "vertical", "vertical.xml")) as f:
+            content = f.read()
+
+        assert reformat_xml(content) == reformat_xml(
+            self.make_test_vertical(
+                solution_as_attribute,
+                # If both are true the expected output should only have the attribute, since it took precedence
+                # and the attribute contents are broken XML
+                solution_as_element if not solution_as_attribute else False,
+            )
+        )
```

Comment:
_sigh_ I typed out this comment before and apparently forgot to add it the the review before submitting. Here goes again...

I know this is being done to make sure the tests are DRY, but the forked logic in these test cases and helper methods makes it hard to read and reason about these tests. Something like this would be more explicit and readable, IMO

```python
    def make_test_vertical(self, solution_attribute=None, solution_element=None):
        """Create a test vertical with an SGA unit inside"""
        solution_attribute = 'solution="{}"'.format(solution_attribute) if solution_attribute else None
        solution_element = '<solution><p>{}</p></solution>'.format(solution_element) if solution_element else None

        return (
            """<vertical display_name="SGA Unit">
              <edx_sga url_name="edx_sga" xblock-family="xblock.v1" display_name="SGA Test 1" {solution_attribute}>
                {solution_element}
              </edx_sga>
            </vertical>""".format(solution_attribute=solution_attribute, solution_element=solution_element)
        )
    
    def import_test_course(self, solution_attribute=None, solution_element=None):
        # ...

    @data(*[
        ['attribute value', 'element value', 'attribute value'],
        ['attribute value', None, 'attribute value'],
        [None, 'element value', 'element value'],
        [None, None, ''],
    ])
    @unpack
    def test_import(self, solution_attribute_value, solution_element_value, expected_solution_text):
        """Import the test course with the SGA module"""
        course = self.import_test_course(solution_attribute_value, solution_element_value)
        sga = course.get_children()[0].get_children()[0].get_children()[0].get_children()[0]
        assert expected_solution_text in sga.solution
        assert sga.showanswer == ShowAnswer.PAST_DUE
```

Reply:
Sounds good